client/windows/CMapOverview.cpp

```diff
@@ -0,0 +1,204 @@
+#ifdef VCMI_MOBILE
+	moveBy({0, -pos.h / 2});
+#endif
+	fitToScreen(10);
+}
+
+Canvas CMapOverview::CMapOverviewWidget::createMinimapForLayer(std::unique_ptr<CMap> & map, int layer) const
+{
+	Canvas canvas = Canvas(Point(map->width, map->height));
+
+	for (int y = 0; y < map->height; ++y)
+		for (int x = 0; x < map->width; ++x)
+		{
+			TerrainTile & tile = map->getTile(int3(x, y, layer));
+
+			ColorRGBA color = tile.terType->minimapUnblocked;
+			if (tile.blocked && (!tile.visitable))
+				color = tile.terType->minimapBlocked;
+
+			if(drawPlayerElements)
+				// if object at tile is owned - it will be colored as its owner
+				for (const CGObjectInstance *obj : tile.blockingObjects)
+				{
+					PlayerColor player = obj->getOwner();
+					if(player == PlayerColor::NEUTRAL)
+					{
+						color = graphics->neutralColor;
+						break;
+					}
+					if (player.isValidPlayer())
+					{
+						color = graphics->playerColors[player.getNum()];
+						break;
+					}
+				}
+
+			canvas.drawPoint(Point(x, y), color);
+		}
+	
+	return canvas;
+}
+
+std::vector<Canvas> CMapOverview::CMapOverviewWidget::createMinimaps(ResourcePath resource) const
+{
+	std::vector<Canvas> ret = std::vector<Canvas>();
+
+	CMapService mapService;
+	std::unique_ptr<CMap> map;
+	try
+	{
+		map = mapService.loadMap(resource);
+	}
+	catch (...)
+	{
+		return ret;
+	}
+
+	return createMinimaps(map);
+}
+
+std::vector<Canvas> CMapOverview::CMapOverviewWidget::createMinimaps(std::unique_ptr<CMap> & map) const
+{
+	std::vector<Canvas> ret = std::vector<Canvas>();
+
+	for(int i = 0; i < (map->twoLevel ? 2 : 1); i++)
+		ret.push_back(createMinimapForLayer(map, i));
+
+	return ret;
+}
+
+std::shared_ptr<CPicture> CMapOverview::CMapOverviewWidget::buildDrawMinimap(const JsonNode & config) const
+{
+	logGlobal->debug("Building widget drawMinimap");
+
+	auto rect = readRect(config["rect"]);
+	auto id = config["id"].Integer();
+
+	if(id >= minimaps.size())
+		return nullptr;
+
+	Canvas canvasScaled = Canvas(Point(rect.w, rect.h));
+	canvasScaled.drawScaled(minimaps[id], Point(0, 0), Point(rect.w, rect.h));
+	std::shared_ptr<IImage> img = GH.renderHandler().createImage(canvasScaled.getInternalSurface());
+
+	return std::make_shared<CPicture>(img, Point(rect.x, rect.y));
+}
+
+CMapOverview::CMapOverviewWidget::CMapOverviewWidget(CMapOverview& parent):
+	InterfaceObjectConfigurable(), p(parent)
+{
+	drawPlayerElements = p.tabType == ESelectionScreen::newGame;
+
+	const JsonNode config(JsonPath::builtin("config/widgets/mapOverview.json"));
+
+	if(settings["lobby"]["mapPreview"].Bool())
+	{
+		ResourcePath res = ResourcePath(p.resource.getName(), EResType::MAP);
+		std::unique_ptr<CMap> campaignMap = nullptr;
+		if(p.tabType != ESelectionScreen::newGame && config["variables"]["mapPreviewForSaves"].Bool())
+		{
+			CLoadFile lf(*CResourceHandler::get()->getResourceName(ResourcePath(p.resource.getName(), EResType::SAVEGAME)), MINIMAL_SERIALIZATION_VERSION);
```

Comment:
Is this thing fast enough, especially on low end systems, e.g. mobiles?

Reply:
On my 2019 android device it's not noticeable.